bundles/org.openhab.binding.gpio/src/main/java/org/openhab/binding/gpio/internal/handler/PigpioRemoteHandler.java

```diff
@@ -57,60 +63,368 @@ public class PigpioRemoteHandler extends BaseThingHandler {
+                    this.channelHandlers.put(channelUID, new PigpioDigitalInputHandler(configuration, scheduler,
                             state -> updateState(channelUID.getId(), state)));
                 } else if (CHANNEL_TYPE_DIGITAL_OUTPUT.equals(type)) {
                     GPIOOutputConfiguration configuration = channel.getConfiguration()
                             .as(GPIOOutputConfiguration.class);
-                    channelHandlers.put(channelUID, new PigpioDigitalOutputHandler(configuration, jPigpio,
-                            state -> updateState(channelUID.getId(), state)));
+                    PigpioDigitalOutputHandler handler = new PigpioDigitalOutputHandler(configuration, scheduler,
+                            state -> updateState(channelUID.getId(), state));
+                    this.channelHandlers.put(channelUID, handler);
                 }
             } catch (PigpioException e) {
-                logger.warn("Failed to initialize {}: {}", channelUID, e.getMessage());
-            } catch (InvalidPullUpDownException e) {
-                logger.warn("Failed to initialize {}: Invalid Pull Up/Down resistor configuration", channelUID);
-            } catch (NoGpioIdException e) {
-                logger.warn("Failed to initialize {}: GpioId is not set", channelUID);
+                logger.warn("Failed to initialize channel {} {}", channelUID, e.toString());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                        String.format("Failed to initialize channel {} {}", channelUID, e.getLocalizedMessage()));
+            } catch (ChannelConfigurationException e) {
+                logger.error("Invalid configuration for channel {} {}", channelUID, e.toString());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                        String.format("Invalid configuration for channel {} {}", channelUID, e.getLocalizedMessage()));
             }
         });
+
+        logger.debug("gpio channels initialized");
+    }
+
+    protected void setChannelJPigpio(@Nullable JPigpio jPigpio) throws PigpioException {
+        if (this.channelHandlers.isEmpty()) {
+            createChannelHandlers();
+        }
+
+        for (ChannelHandler handler : this.channelHandlers.values()) {
+            handler.listen(jPigpio);
+        }
+
+        logger.debug("gpio jPigpio listening");
+    }
+
+    @Nullable
+    private Future<?> connectionJob = null;
+    /**
+     * Syncronizes all socket related code
+     * to avoid racing.
+     */
+    private Object connectionLock = new Object();
+
+    protected void killConnectionPoll() {
+        if (this.connectionJob != null) {
+            synchronized (this.connectionLock) {
+                if (this.connectionJob != null) {
+                    Future<?> job = this.connectionJob;
+                    this.connectionJob = null;
```

Comment:
move this line to after the if block.

Reply:
I am not sure which line you are referring to here. The race checks here are probably not needed, the only call to this is in Dispose() which is already holding the lock.

Reply:
Not sure what I meant, forgot. Looks good now.